File: WinPaletter/ColorItemControllers/SubMenu.vb
﻿Imports WinPaletter.NativeMethods

Public Class SubMenu

    Private _shown As Boolean
    Private _overrideColor As Color
    Private _eventDone As Boolean
    ReadOnly _Speed As Integer = 20
    ReadOnly _dark As Single = 0.7
    Private PreviousClr As Color

#Region "Form Shadow"

    Private aeroEnabled As Boolean

    Protected Overrides ReadOnly Property CreateParams() As CreateParams
        Get
            CheckAeroEnabled()
            Dim cp As CreateParams = MyBase.CreateParams
            If Not aeroEnabled Then
                cp.ClassStyle = cp.ClassStyle Or Dwmapi.CS_DROPSHADOW
                cp.ExStyle = cp.ExStyle Or 33554432
                Return cp
            Else
                Return cp
            End If
        End Get
    End Property

    Protected Overrides Sub WndProc(ByRef m As Message)
        Select Case m.Msg
            Case Dwmapi.WM_NCPAINT
                Dim val = 2
                If aeroEnabled Then
                    Dwmapi.DwmSetWindowAttribute(Handle, If(GetRoundedCorners(), 2, 1), val, 4)
                    Dim bla As New Dwmapi.MARGINS()
                    With bla
                        .bottomHeight = 1
                        .leftWidth = 1
                        .rightWidth = 1
                        .topHeight = 1
                    End With
                    Dwmapi.DwmExtendFrameIntoClientArea(Handle, bla)
                End If
                Exit Select
        End Select

        Const WM_NCACTIVATE As UInt32 = &H86

        If m.Msg = WM_NCACTIVATE AndAlso m.WParam.ToInt32() = 0 Then
            HandleDeactivate()
        End If

        MyBase.WndProc(m)
    End Sub

    Private Sub CheckAeroEnabled()
        If Environment.OSVersion.Version.Major >= 6 Then
            Dim Com As Boolean
            Dwmapi.DwmIsCompositionEnabled(Com)
            aeroEnabled = Com
        Else
            aeroEnabled = False
        End If
    End Sub
#End Region


    Public Function ShowMenu(ColorItem As UI.Controllers.ColorItem, Optional EnableDelete As Boolean = False) As Color
        Button5.Visible = EnableDelete

        MainColor.BackColor = ColorItem.BackColor.CB((Trackbar1.Value - 100) / 100)
        DefaultColor.BackColor = ColorItem.DefaultColor.CB((Trackbar2.Value - 100) / 100)
        InvertedColor.BackColor = ColorItem.BackColor.Invert.CB((Trackbar3.Value - 100) / 100)

        MainColor.DefaultColor = ColorItem.BackColor
        DefaultColor.DefaultColor = ColorItem.DefaultColor
        InvertedColor.DefaultColor = ColorItem.BackColor.Invert

        If ColorItem.ColorsHistory.Count > 1 Then
            PreviousColor.BackColor = ColorItem.ColorsHistory.Item(ColorItem.ColorsHistory.Count - 2).CB((Trackbar4.Value - 100) / 100)
            PreviousColor.DefaultColor = ColorItem.ColorsHistory.Item(ColorItem.ColorsHistory.Count - 2)
        Else
            PreviousColor.BackColor = ColorItem.BackColor.CB((Trackbar4.Value - 100) / 100)
            PreviousColor.DefaultColor = ColorItem.BackColor
        End If

        PreviousClr = PreviousColor.DefaultColor

        GetHistoryColors(ColorItem)

        If ShowDialog() = DialogResult.OK Then
            Select Case My.Application.ColorEvent
                Case My.MyApplication.MenuEvent.Copy
                    My.Application.CopiedColor = MainColor.BackColor
                    Return ColorItem.BackColor

                Case My.MyApplication.MenuEvent.Cut
                    My.Application.CopiedColor = MainColor.BackColor
                    ColorItem.BackColor = Color.Black

                Case My.MyApplication.MenuEvent.Paste
                    ColorItem.BackColor = My.Application.CopiedColor
                    Return My.Application.CopiedColor

                Case My.MyApplication.MenuEvent.Override
                    ColorItem.BackColor = _overrideColor
                    Return ColorItem.BackColor

                Case My.MyApplication.MenuEvent.Delete
                    ColorItem.BackColor = Color.FromArgb(0, 0, 0, 0)
                    Return ColorItem.BackColor

                Case My.MyApplication.MenuEvent.None
                    Return MainColor.DefaultColor

            End Select
        Else
            My.Application.ColorEvent = My.MyApplication.MenuEvent.None
            Return MainColor.DefaultColor 'Nothing
        End If
    End Function

    Private Sub SubMenu_Shown(sender As Object, e As EventArgs) Handles Me.Shown
        _shown = True

        PaletteContainer.Visible = False
    End Sub

    Private Sub SubMenu_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        User32.AnimateWindow(Handle, _Speed, User32.AnimateWindowFlags.AW_HIDE Or User32.AnimateWindowFlags.AW_BLEND)
    End Sub

    Sub HandleDeactivate()
        If _shown Then
            _shown = False
            If Not _eventDone Then DialogResult = DialogResult.None
            Me.Close()
        End If
    End Sub

    Sub Collapse_Expand()

        Button4.Visible = False

        Select Case PaletteContainer.Visible
            Case False
                Button4.Text = "<"

                For i = PaletteContainer.Left + 3 To PaletteContainer.Right + 8 Step 2
                    Width = i
                Next

                Width = PaletteContainer.Right + 8

                PaletteContainer.Visible = True
                Label2.Visible = True

            Case True
                Button4.Text = ">"

                PaletteContainer.Visible = False
                Label2.Visible = False

                For i = PaletteContainer.Right + 8 To PaletteContainer.Left + 3 Step -2
                    Width = i
                Next

                Width = PaletteContainer.Left + 3

        End Select

        Button4.Visible = True
    End Sub

    Private Sub SubMenu_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        _shown = False

        Dim p As Point = MousePosition

        If p.Y + Height > My.Computer.Screen.WorkingArea.Bottom Then
            p.Y = My.Computer.Screen.WorkingArea.Bottom - Height - 5
        End If

        If p.Y < My.Computer.Screen.WorkingArea.Top Then
            p.Y = 0
        End If

        If p.X + Width > My.Computer.Screen.WorkingArea.Right Then
            p.X = My.Computer.Screen.WorkingArea.Right - Width - 5
        End If

        If p.X < My.Computer.Screen.WorkingArea.Left Then
            p.X = 0
        End If


        Location = p

        Width = PaletteContainer.Left + 3

        PaletteContainer.Visible = False
        Label2.Visible = False
        Button4.Text = ">"

        LoadLanguage
        ApplyStyle(Me)

        If My.Application.CopiedColor = Nothing Then

            Button3.Enabled = False

            Try
                If Clipboard.GetData("Text") IsNot Nothing Then
                    Dim s As String = Clipboard.GetData("Text").ToString.ToLower

                    If s.StartsWith("color ") Then
                        Dim C As Color = Color.FromArgb(255, 0, 0, 0)
                        s = s.Remove(0, "color ".Count)
                        s = s.Replace("[", "")
                        s = s.Replace("]", "")
                        s = s.Replace(" ", "")

                        For Each x As String In s.Split(",")
                            Dim i As Integer = Val(x.Remove(0, 2))
                            If x.StartsWith("a=") Then C = Color.FromArgb(i, C)
                            If x.StartsWith("r=") Then C = Color.FromArgb(C.A, i, C.G, C.B)
                            If x.StartsWith("g=") Then C = Color.FromArgb(C.A, C.R, i, C.B)
                            If x.StartsWith("b=") Then C = Color.FromArgb(C.A, C.R, C.G, i)
                        Next

                        My.Application.CopiedColor = C
                        Button3.Enabled = True
                    End If
                End If

            Catch
                Button3.Enabled = False
            End Try

        Else
            Button3.Enabled = True
        End If

        BackColor = If(My.Style.DarkMode, MainColor.BackColor.Dark(_dark), MainColor.BackColor.LightLight)

        User32.AnimateWindow(Handle, _Speed, User32.AnimateWindowFlags.AW_ACTIVATE Or User32.AnimateWindowFlags.AW_BLEND)

        Invalidate()
    End Sub

    Sub MiniColorItem_Clicked(sender As Object, e As EventArgs)
        MainColor.BackColor = sender.BackColor
        MainColor.DefaultColor = sender.BackColor

        InvertedColor.BackColor = MainColor.BackColor.Invert.CB((Trackbar3.Value - 100) / 100)
        InvertedColor.DefaultColor = MainColor.BackColor.Invert

        Collapse_Expand()
    End Sub

    Sub GetHistoryColors(ColorItem As UI.Controllers.ColorItem)
        PaletteContainer.SuspendLayout()

        For Each c As UI.Controllers.ColorItem In PaletteContainer.Controls.OfType(Of UI.Controllers.ColorItem)
            RemoveHandler c.Click, AddressOf MiniColorItem_Clicked
            c.Dispose()
            PaletteContainer.Controls.Remove(c)
        Next

        PaletteContainer.Controls.Clear()

        For Each c As Color In ColorItem.ColorsHistory
            Dim MiniColorItem As New UI.Controllers.ColorItem With {
                .Size = .GetMiniColorItemSize,
                .AllowDrop = False,
                .PauseColorsHistory = True,
                .BackColor = c, .DefaultColor = .BackColor}

            PaletteContainer.Controls.Add(MiniColorItem)
            AddHandler MiniColorItem.Click, AddressOf MiniColorItem_Clicked
        Next

        PaletteContainer.ResumeLayout()
    End Sub

    Sub Update_Variants()
        InvertedColor.DefaultColor = MainColor.DefaultColor.Invert
    End Sub

    Private Sub Button1_Click(sender As Object, e As EventArgs) Handles Button1.Click
        Clipboard.SetData("Text", MainColor.BackColor)
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Copy
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub Button2_Click(sender As Object, e As EventArgs) Handles Button2.Click
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Cut
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub Button3_Click(sender As Object, e As EventArgs) Handles Button3.Click
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Paste
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub MainColor_Click(sender As Object, e As EventArgs) Handles MainColor.Click, InvertedColor.Click, DefaultColor.Click
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Override
        _overrideColor = sender.BackColor
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub Button4_Click(sender As Object, e As EventArgs) Handles Button4.Click
        Collapse_Expand()
    End Sub

    Private Sub Button5_Click(sender As Object, e As EventArgs) Handles Button5.Click
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Delete
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub Trackbar1_Scroll(sender As Object) Handles Trackbar1.Scroll
        MainColor.BackColor = MainColor.DefaultColor.CB((Trackbar1.Value - 100) / 100)
    End Sub

    Private Sub Trackbar2_Scroll(sender As Object) Handles Trackbar2.Scroll
        DefaultColor.BackColor = DefaultColor.DefaultColor.CB((Trackbar2.Value - 100) / 100)
    End Sub

    Private Sub Trackbar3_Scroll(sender As Object) Handles Trackbar3.Scroll
        InvertedColor.BackColor = InvertedColor.DefaultColor.CB((Trackbar3.Value - 100) / 100)
    End Sub

    Private Sub Button6_Click(sender As Object, e As EventArgs) Handles Button6.Click
        Trackbar1.Value = 100
    End Sub

    Private Sub Button7_Click(sender As Object, e As EventArgs) Handles Button7.Click
        Trackbar2.Value = 100
    End Sub

    Private Sub Button8_Click(sender As Object, e As EventArgs) Handles Button8.Click
        Trackbar3.Value = 100
    End Sub

    Private Sub Trackbar4_Scroll(sender As Object) Handles Trackbar4.Scroll
        PreviousColor.BackColor = PreviousColor.DefaultColor.CB((Trackbar4.Value - 100) / 100)
    End Sub

    Private Sub Button9_Click(sender As Object, e As EventArgs) Handles Button9.Click
        Trackbar4.Value = 100
    End Sub

    Private Sub PreviousColor_Click(sender As Object, e As EventArgs) Handles PreviousColor.Click
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Override
        _overrideColor = sender.BackColor
        DialogResult = DialogResult.OK
        Close()
    End Sub

    Private Sub Button10_Click(sender As Object, e As EventArgs) Handles Button10.Click
        _eventDone = True
        My.Application.ColorEvent = My.MyApplication.MenuEvent.Override
        _overrideColor = PreviousClr
        DialogResult = DialogResult.OK
        Close()
    End Sub
End Class

File: WinPaletter/ColorItemControllers/SubMenu.cs
using FluentTransitions;
using Microsoft.VisualBasic;
using Serilog.Sinks.File;
using System;
using System.Drawing;
using System.Linq;
using System.Windows.Forms;
using WinPaletter.UI.Controllers;

namespace WinPaletter
{
    public partial class SubMenu
    {
        private Color _overrideColor;
        private bool _eventDone;
        private readonly float _dark = 0.7f;
        private Color PreviousClr;

        public SubMenu()
        {
            InitializeComponent();
        }

        public Color ShowMenu(ColorItem ColorItem, bool EnableDelete = false)
        {
            Button5.Visible = EnableDelete;

            MainColor.BackColor = ColorItem.BackColor.CB((float)((trackBarX1.Value - 100) / 100d));
            DefaultColor.BackColor = ColorItem.DefaultBackColor.CB((float)((trackBarX2.Value - 100) / 100d));
            InvertedColor.BackColor = ColorItem.BackColor.Invert().CB((float)((trackBarX4.Value - 100) / 100d));

            MainColor.DefaultBackColor = ColorItem.BackColor;
            DefaultColor.DefaultBackColor = ColorItem.DefaultBackColor;
            InvertedColor.DefaultBackColor = ColorItem.BackColor.Invert();

            if (ColorItem.ColorsHistory.Count > 1)
            {
                PreviousColor.BackColor = ColorItem.ColorsHistory[ColorItem.ColorsHistory.Count - 2].CB((float)((trackBarX3.Value - 100) / 100d));
                PreviousColor.DefaultBackColor = ColorItem.ColorsHistory[ColorItem.ColorsHistory.Count - 2];
            }
            else
            {
                PreviousColor.BackColor = ColorItem.BackColor.CB((float)((trackBarX3.Value - 100) / 100d));
                PreviousColor.DefaultBackColor = ColorItem.BackColor;
            }

            PreviousClr = PreviousColor.DefaultBackColor;

            GetHistoryColors(ColorItem);

            if (ShowDialog() == DialogResult.OK)
            {
                switch (ColorClipboard.Event)
                {
                    case ColorClipboard.MenuEvent.Copy:
                        {
                            ColorClipboard.CopiedColor = MainColor.BackColor;
                            return ColorItem.BackColor;
                        }

                    case ColorClipboard.MenuEvent.Cut:
                        {
                            ColorClipboard.CopiedColor = MainColor.BackColor;
                            ColorItem.BackColor = Color.Black;
                            break;
                        }

                    case ColorClipboard.MenuEvent.Paste:
                        {
                            ColorItem.BackColor = ColorClipboard.CopiedColor;
                            return ColorClipboard.CopiedColor;
                        }

                    case ColorClipboard.MenuEvent.Override:
                        {
                            ColorItem.BackColor = _overrideColor;
                            return ColorItem.BackColor;
                        }

                    case ColorClipboard.MenuEvent.Delete:
                        {
                            ColorItem.BackColor = Color.FromArgb(0, 0, 0, 0);
                            return ColorItem.BackColor;
                        }

                    case ColorClipboard.MenuEvent.None:
                        {
                            return MainColor.DefaultBackColor;
                        }

                }
            }
            else
            {
                ColorClipboard.Event = ColorClipboard.MenuEvent.None;
                return MainColor.DefaultBackColor;
            } // Nothing

            return default;
        }

        private void SubMenu_Shown(object sender, EventArgs e)
        {
            PaletteContainer.Visible = false;
        }

        public void Collapse_Expand()
        {
            Button4.Visible = false;

            if (PaletteContainer.Visible)
            {
                Button4.Text = ">";

                Transition
                    .With(this, nameof(Width), PaletteContainer.Left + 3)
                    .CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration));

                PaletteContainer.Visible = false;
                Label2.Visible = false;
            }
            else
            {
                Button4.Text = "<";

                Transition
                    .With(this, nameof(Width), PaletteContainer.Right + 8)
                    .CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration));

                PaletteContainer.Visible = true;
                Label2.Visible = true;
            }

            Button4.Visible = true;
        }

        private void SubMenu_Load(object sender, EventArgs e)
        {
            Icon = FormsExtensions.Icon<MainForm>();

            Point p = MousePosition;

            if (p.Y + Height > Screen.PrimaryScreen.WorkingArea.Bottom)
            {
                p.Y = Screen.PrimaryScreen.WorkingArea.Bottom - Height - 5;
            }

            if (p.Y < Screen.PrimaryScreen.WorkingArea.Top)
            {
                p.Y = 0;
            }

            if (p.X + Width > Screen.PrimaryScreen.WorkingArea.Right)
            {
                p.X = Screen.PrimaryScreen.WorkingArea.Right - Width - 5;
            }

            if (p.X < Screen.PrimaryScreen.WorkingArea.Left)
            {
                p.X = 0;
            }

            Location = p;

            Width = PaletteContainer.Left + 3;

            PaletteContainer.Visible = false;
            Label2.Visible = false;
            Button4.Text = ">";

            if (Clipboard.GetData("Text") is not null)
            {
                string s = Clipboard.GetData("Text").ToString().ToLower();

                if (s.StartsWith("color "))
                {
                    Color C = Color.FromArgb(255, 0, 0, 0);
                    s = s.Remove(0, "color ".Count());
                    s = s.Replace("[", string.Empty);
                    s = s.Replace("]", string.Empty);
                    s = s.Replace(" ", string.Empty);

                    foreach (string x in s.Split(','))
                    {
                        int i = (int)Math.Round(Conversion.Val(x.Remove(0, 2)));
                        if (x.StartsWith("a="))
                            C = Color.FromArgb(i, C);
                        if (x.StartsWith("r="))
                            C = Color.FromArgb(C.A, i, C.G, C.B);
                        if (x.StartsWith("g="))
                            C = Color.FromArgb(C.A, C.R, i, C.B);
                        if (x.StartsWith("b="))
                            C = Color.FromArgb(C.A, C.R, C.G, i);
                    }

                    ColorClipboard.CopiedColor = C;
                }

                else if (s.IsHexColor())
                {
                    ColorClipboard.CopiedColor = s.ToColorFromHex();
                }
            }

            Button3.Enabled = ColorClipboard.CopiedColor != Color.Empty;

            BackColor = Program.Style.DarkMode ? MainColor.BackColor.Dark(_dark) : MainColor.BackColor.LightLight();
        }

        public void MiniColorItem_Clicked(object sender, EventArgs e)
        {
            MainColor.BackColor = ((ColorItem)sender).BackColor;
            MainColor.DefaultBackColor = ((ColorItem)sender).BackColor;

            InvertedColor.BackColor = MainColor.BackColor.Invert().CB((float)((trackBarX4.Value - 100) / 100d));
            InvertedColor.DefaultBackColor = MainColor.BackColor.Invert();

            Collapse_Expand();
        }

        public void GetHistoryColors(ColorItem ColorItem)
        {
            PaletteContainer.SuspendLayout();

            foreach (UI.Controllers.ColorItem c in PaletteContainer.Controls.OfType<UI.Controllers.ColorItem>())
            {
                c.Click -= MiniColorItem_Clicked;
                c.Dispose();
                PaletteContainer.Controls.Remove(c);
            }

            PaletteContainer.Controls.Clear();

            foreach (Color c in ColorItem.ColorsHistory)
            {
                UI.Controllers.ColorItem MiniColorItem = new();
                MiniColorItem.Size = MiniColorItem.GetMiniColorItemSize();
                MiniColorItem.AllowDrop = false;
                MiniColorItem.PauseColorsHistory = true;
                MiniColorItem.BackColor = c;
                MiniColorItem.DefaultBackColor = MiniColorItem.BackColor;

                PaletteContainer.Controls.Add(MiniColorItem);
                MiniColorItem.Click += MiniColorItem_Clicked;
            }

            PaletteContainer.ResumeLayout();
        }

        public void Update_Variants()
        {
            InvertedColor.DefaultBackColor = MainColor.DefaultBackColor.Invert();
        }

        private void Button1_Click(object sender, EventArgs e)
        {
            Clipboard.SetData("Text", MainColor.BackColor.ToStringHex(true, true));
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Copy;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void Button2_Click(object sender, EventArgs e)
        {
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Cut;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void Button3_Click(object sender, EventArgs e)
        {
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Paste;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void MainColor_Click(object sender, EventArgs e)
        {
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Override;
            _overrideColor = ((ColorItem)sender).BackColor;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void Button4_Click(object sender, EventArgs e)
        {
            Collapse_Expand();
        }

        private void Button5_Click(object sender, EventArgs e)
        {
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Delete;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void PreviousColor_Click(object sender, EventArgs e)
        {
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Override;
            _overrideColor = ((ColorItem)sender).BackColor;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void Button10_Click(object sender, EventArgs e)
        {
            _eventDone = true;
            ColorClipboard.Event = ColorClipboard.MenuEvent.Override;
            _overrideColor = PreviousClr;
            DialogResult = DialogResult.OK;
            Close();
        }

        private void trackBarX1_ValueChanged(object sender, EventArgs e)
        {
            MainColor.BackColor = MainColor.DefaultBackColor.CB((float)((trackBarX1.Value - 100) / 100d));
        }

        private void trackBarX2_ValueChanged(object sender, EventArgs e)
        {
            DefaultColor.BackColor = DefaultColor.DefaultBackColor.CB((float)((trackBarX2.Value - 100) / 100d));
        }

        private void trackBarX3_ValueChanged(object sender, EventArgs e)
        {
            PreviousColor.BackColor = PreviousColor.DefaultBackColor.CB((float)((trackBarX3.Value - 100) / 100d));
        }

        private void trackBarX4_ValueChanged(object sender, EventArgs e)
        {
            InvertedColor.BackColor = InvertedColor.DefaultBackColor.CB((float)((trackBarX4.Value - 100) / 100d));
        }
    }
}